src/core/reaction_ensemble.hpp

```diff
@@ -389,9 +389,64 @@ class WidomInsertion : public ReactionAlgorithm {
   std::pair<double, double> measure_excess_chemical_potential(int reaction_id);
 };
 
-//////////////////////////////////////////////////////////////////free functions
+///////////////////////
+// utility functions //
+///////////////////////
+
 double calculate_factorial_expression(SingleReaction &current_reaction,
                                       std::map<int, int> &old_particle_numbers);
+
 double factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i(int Ni0, int nu_i);
+
+/**
+ * Calculate the average of an array (used for the histogram of the
+ * Wang-Landau algorithm). It excludes values which are initialized to be
+ * negative. Those values indicate that the Wang-Landau algorithm should not
+ * sample those values. The values still occur in the list because we can only
+ * store "rectangular" value ranges.
+ */
+template <typename T>
+double average_list_of_allowed_entries(const std::vector<T> &rng) {
+  T result = 0;
+  int counter_allowed_entries = 0;
+  for (auto &val : rng) {
+    if (val >= 0) { // checks for validity of index i (think of energy
+                    // collective variables, in a cubic memory layout
+                    // there will be indices which are not allowed by
+                    // the energy boundaries. These values will be
+                    // initialized with a negative fill value)
+      result += val;
+      counter_allowed_entries += 1;
+    }
+  }
+  if (counter_allowed_entries) {
+    return static_cast<double>(result) / counter_allowed_entries;
+  }
+  return 0.0;
+}
+
+/**
+ * Checks whether a number is in an array.
+ */
+inline bool is_in_vector(int value, std::vector<int> const &rng) {
+  return boost::range::find(rng, value) != rng.end();
+}
+
+/**
+ * Finds the minimum non negative value in the provided double array and returns
+ * this value.
+ */
+inline double find_minimum_non_negative_value(std::vector<double> const &rng) {
```

Comment:
This should probably be formulated as an accumulate. This implementation is UB if `rng` is empty, which needs to be checked.